### PR TITLE
Make Suggested Actions accessible again!

### LINKED ISSUE #1780: Suggested actions are not spoken
Given the server returns a reply that has SuggestedActions with Speak property populated. The WebChat wouldn't speak the given reply by the server.

### LINKED ISSUE #2277: Carousel Layout text activities are repeated multiple times by AT due to nesting <Bubble>
## To Reproduce

1. Open Mock Bot
2. Turn on Narrator (Edge) or NVDA (Chrome)
3. Type 'layout carousel' into Mock Bot
4. Note that the text activities (non-Markdown) are read multiple times

## Expected behavior

Text activities in carousels should only be read once. This is caused by the nesting of <Bubble> in `<CarouselLayout>`.

## Additional context
Uncovered from issue #2273

[Bug]

Note to fixer:
also test Mock Bot command `layout`

### LINKED ISSUE #2285: Improvements to Web Chat a11y
## Feature Request

**Describe the suggestion or request in detail**

From my recent work on a11y, I have suggestions I want to broach on small improvements to make on AT experience in Web Chat.

- `aria-label` for both the Transcript (role of log) and Connection status (role of status) section of the app

I will keep adding brainstorming ideas here as I receive them / think of them.

[Enhancement]

### PR DESCRIPTION
**Pull request by corinagum**
(jk they were never accessible) 

Fixes #1780
Fixes #2277
Fixes #2285

## Changelog Entry

- Fixes [#1780](https://github.com/microsoft/BotFramework-WebChat/issues/1780), [#2277](https://github.com/microsoft/BotFramework-WebChat/issues/2277), and [#2285](https://github.com/microsoft/BotFramework-WebChat/issues/2285). Make Suggested Actions accessible, Fix Markdown card in carousel being read multiple times, and label widgets of Connectivity Status and Suggested Actions containers, by [@corinagum](https://github.com/corinagum) in PR [#2613](https://github.com/microsoft/BotFramework-WebChat/pull/2613)

### Note: How to test Suggested Actions container on Mock Bot
- use command `emptycard` or
- use command `suggested-actions`

## Description
- Make Suggested Actions container accessible. See Browser Discrepancies below.
- Fix Markdown card in carousel being read multiple times - previously I recorded this as a nesting issue but was actually a problem with the attachment types inside. This has been resolved with `aria-live=" "` and `aria-hidden={true}` inside the Carousel component
- Connectivity Status: AT will now read "Connectivity Status: Connected/Unable to  connect/etc" when it reaches the (Connectivity Status container) - in the UI this component only shows when not connected. Changes in connectivity status will also be announced by the AT, since this is a status widget.
- Suggested Actions container: This component is hidden until buttons are added inside. At such time, the AT will announce "Suggested Actions container: has content" regardless of where the user is focused on the app, due to this component being a status widget. When the Suggested Actions container is empty, AT will announce "Suggested Actions container: is empty"
-  Japanese localization file updated to latest
- Please note, "Upload file text" in localization files has been modified. See `en-us` for changes.

### Browser discrepancies 

Previously, Suggested Actions were ignored by the AT. This PR fixes that, HOWEVER, there are some idiosyncrasies across browsers that are documented below:

1. On Chrome & NVDA: When the Suggested Actions container changes, it will read "Suggested actions container: has content. Blue Red Green" twice. I've gone over and over the code with @compulim as my sanity check and I'm confident that this is caused by a Chrome error, not Web Chat's code. 
1. Firefox & NVDA: Throughout the app, the the AT will randomly read extra letters that are not in the aria-label or visual on the UI. For example, it will say "Suggested Actions: has content. *ED*". 
1. Edge & Narrator: AT will read the change in Suggested Actions container "Suggested Actions: has content. Contains three items", but it will *NOT* read the buttons inside without navigating to that container. Other browsers will say "Suggestesd Actions container: has content. Blue red Green".
1. Safari & VoiceOver: AT will read the Connectivity Status and SuggestedActions labels/containers multiple times on start up. 

---

-  [ ] Testing Added

### CODE REVIEW

File: packages/component/src/Activity/StackedLayout.js [hunk 1]
```diff
@@ -146,7 +152,7 @@ const StackedLayout = ({ activity, children, timestampClassName }) => {
             <div className="filler" />
           </div>
         )}
-        {/* Because of differences in browser implementations, aria-label=" " is used to make the screen reader not repeat the same text multiple times in Chrome v75 */}
+        {/* Because of differences in browser implementations, aria-label=" " is used to make the screen reader not repeat the same text multiple times in Chrome v75 and Edge 44 */}
```
Comment (corinagum) [inline packages/component/src/Activity/StackedLayout.js]: needs fix

File: packages/component/src/ScreenReaderText.js [hunk 2]
```diff
@@ -2,25 +2,45 @@ import { css } from 'glamor';
 import classNames from 'classnames';
 import PropTypes from 'prop-types';
 import React from 'react';
+import * as browser from '../lib/Utils/detectBrowser';
+// let bgC;
```
Comment (corinagum) [inline packages/component/src/ScreenReaderText.js]: delete

File: packages/component/src/ScreenReaderText.js [hunk 3]
```diff
@@ -2,25 +2,45 @@ import { css } from 'glamor';
...
+  // fontFamily: 'sans-serif',
+  // backgroundColor: bgC,
+  // color: 'black',
+  // padding: '2px',
+  // original:
   color: 'transparent',
   height: 1,
-  left: -10000,
   overflow: 'hidden',
   position: 'absolute',
```
Comment (corinagum) [inline packages/component/src/ScreenReaderText.js]: delete

File: packages/component/src/Activity/StackedLayout.js [hunk 4]
```diff
@@ -117,7 +119,11 @@ const StackedLayout = ({ activity, children, timestampClassName }) => {
   const botAriaLabel = useLocalize('Bot said something', initials, plainText);
   const userAriaLabel = useLocalize('User said something', initials, plainText);
 
-  const ariaLabel = fromUser ? userAriaLabel : botAriaLabel;
+  const sentAtTimestamp = useLocalize('SentAt') + CalculateAbsoluteTime(timestamp);
+
+  const someoneSaidString = (fromUser ? userAriaLabel : botAriaLabel).trim();
```
Comment (tdurnford) [inline packages/component/src/Activity/StackedLayout.js]: Not a huge fan of that variable name

Comment (corinagum) [reply packages/component/src/Activity/StackedLayout.js]: Hmm I suppose, but we have variables in other places saying "BotSaidx" and "UserSaidx", so it's not like there isn't precedent. 

Got other suggestions?

File: packages/component/src/Attachment/TextContent.js [hunk 5]
```diff
@@ -1,10 +1,14 @@
-// The content rendered here is sanitized.
+/* eslint-disable no-sync */
```
Comment (tdurnford) [inline packages/component/src/Attachment/TextContent.js]: Do we need this?

Comment (corinagum) [reply packages/component/src/Attachment/TextContent.js]: yup. see processSync funciton below.

File: packages/core/src/reducers/suggestedActions.js [hunk 6]
```diff
@@ -6,7 +6,11 @@ const DEFAULT_STATE = [];
 export default function suggestedActions(state = DEFAULT_STATE, { payload = {}, type }) {
   switch (type) {
     case SET_SUGGESTED_ACTIONS:
-      state = [].slice.call(payload.suggestedActions || []);
+      if ((payload.suggestedActions || []).length) {
+        state = [].slice.call(payload.suggestedActions || []);
```
Comment (compulim) [inline packages/core/src/reducers/suggestedActions.js]: My bad.

```suggestion
        state = [].slice.call(payload.suggestedActions);
```

File: packages/component/src/Attachment/TextContent.js [hunk 7]
```diff
@@ -26,7 +33,13 @@ const TextContent = ({ contentType, text }) => {
   ) : (
     (text || '').split('\n').map((line, index) => (
       <React.Fragment key={index}>
-        <ScreenReaderText text={text} />
+        <ScreenReaderText
+          text={
+            remark()
```
Comment (tdurnford) [inline packages/component/src/Attachment/TextContent.js]: Is it possible to reuse the `remark` object from earlier? If not, we should probably put this in a variable instead of doing it inline. Also, we should probably memoize this.

File: packages/component/src/Attachment/TextContent.js [hunk 8]
```diff
@@ -13,10 +17,13 @@ import useStyleSet from '../hooks/useStyleSet';
 const TextContent = ({ contentType, text }) => {
   const renderMarkdownAsHTML = useRenderMarkdownAsHTML();
   const [{ textContent: textContentStyleSet }] = useStyleSet();
+  const strippedText = remark()
```
Comment (tdurnford) [inline packages/component/src/Attachment/TextContent.js]: 'useMemo'?

Comment (corinagum) [reply packages/component/src/Attachment/TextContent.js]: This is making use of the packages remark and stripMarkdown -- just following the recommended syntax from there.

File: packages/component/src/Localization/ja-JP.js [hunk 9]
```diff
@@ -41,14 +41,24 @@ function botSaidSomething(avatarInitials, text) {
   return `ボット${avatarInitials}は${text}と言いました`;
 }
 
+function downloadFileWithFileSize(downloadFileText, fileName, size) {
```
Comment (tdurnford) [inline packages/component/src/Localization/ja-JP.js]: I'm just gonna trust you on this file....

Comment (corinagum) [reply packages/component/src/Localization/ja-JP.js]: lmao

File: packages/component/src/Utils/AbsoluteTime.js [hunk 10]
```diff
@@ -6,14 +6,20 @@ import ScreenReaderText from '../ScreenReaderText';
 import useLocalize from '../hooks/useLocalize';
 import useLocalizeDate from '../hooks/useLocalizeDate';
 
-const AbsoluteTime = ({ value }) => {
+const CalculateAbsoluteTime = value => {
```
Comment (tdurnford) [inline packages/component/src/Utils/AbsoluteTime.js]: This seems redundant.... Just use `useLocalizeDate()` in `AbsoluteTime`.

File: packages/component/src/Utils/AbsoluteTime.js [hunk 11]
```diff
@@ -6,14 +6,20 @@ import ScreenReaderText from '../ScreenReaderText';
...
   const localizedTime = useLocalizeDate(value);
-  const text = useLocalize('SentAt') + localizedTime;
+
+  return localizedTime;
+};
+
+const AbsoluteTime = ({ value }) => {
+  const absTime = CalculateAbsoluteTime(value);
```
Comment (tdurnford) [inline packages/component/src/Utils/AbsoluteTime.js]:
```suggestion
  const absTime = useLocalizeDate(value);
```

File: packages/component/src/Utils/detectBrowser.js [hunk 12]
```diff
@@ -0,0 +1,10 @@
+const { userAgent } = navigator;
+
+// not DRY so that list can be alphabetical
```
Comment (tdurnford) [inline packages/component/src/Utils/detectBrowser.js]: Huh?

Comment (corinagum) [reply packages/component/src/Utils/detectBrowser.js]: Ignore, you can't read regex. xD

File: packages/component/src/Utils/AbsoluteTime.js (hunk 11 above)
Comment (corinagum) [reply packages/component/src/Utils/AbsoluteTime.js]: These are different, lol.

File: samples/17.b.clear-after-idle/package-lock.json [hunk 13]
```diff
@@ -3381,6 +3381,7 @@
             "minipass": {
```
Comment (tdurnford) [inline samples/17.b.clear-after-idle/package-lock.json]: Why is the sample package-lock being change?

File: packages/component/src/Utils/AbsoluteTime.js (hunk 10 above)
Comment (corinagum) [reply packages/component/src/Utils/AbsoluteTime.js]: It's not. I had to extract CalculateAbsoluteTime because it's being used in other places.

File: samples/17.b.clear-after-idle/package-lock.json (hunk 13 above)
Comment (corinagum) [reply samples/17.b.clear-after-idle/package-lock.json]: I think I did npm audit a while back and this popped up.

File: packages/core/src/reducers/suggestedActions.js [hunk 14]
```diff
@@ -6,7 +6,11 @@ const DEFAULT_STATE = [];
...
+      } else {
+        state = DEFAULT_STATE;
```
Comment (tdurnford) [inline packages/core/src/reducers/suggestedActions.js]: Why is this necessary? It should already be in it's default state since it's reset before every incoming activity.

File: packages/component/src/Utils/AbsoluteTime.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (corinagum) [reply packages/component/src/Utils/AbsoluteTime.js]: woops

File: packages/core/src/reducers/suggestedActions.js (hunk 14 above)
Comment (corinagum) [reply packages/core/src/reducers/suggestedActions.js]: Accessibility issue, it was rerendering every time even though technically the content isn't changed.

File: packages/component/src/Activity/StackedLayout.js (hunk 4 above)
Comment (tdurnford) [reply packages/component/src/Activity/StackedLayout.js]: Lol no

File: packages/component/src/Utils/detectBrowser.js (hunk 12 above)
Comment (tdurnford) [reply packages/component/src/Utils/detectBrowser.js]: I was referring to the comment. 'not DRY'?

File: packages/component/src/Utils/AbsoluteTime.js (hunk 11 above)
Comment (tdurnford) [reply packages/component/src/Utils/AbsoluteTime.js]: How are they different? William recommended the same thing.

File: packages/component/src/Attachment/TextContent.js (hunk 7 above)
Comment (corinagum) [reply packages/component/src/Attachment/TextContent.js]: yes on the remark, I'll make a util. no on the memoize. xD

File: packages/component/src/Utils/AbsoluteTime.js (hunk 10 above)
Comment (tdurnford) [reply packages/component/src/Utils/AbsoluteTime.js]: Just use `useLocalizeDate` in those other places.
